cpp/ql/lib/semmle/code/cpp/rangeanalysis/SimpleRangeAnalysis.qll

```diff
@@ -1530,6 +1530,22 @@ private predicate isUnsupportedGuardPhi(Variable v, RangeSsaDefinition phi, Vari
   )
 }
 
+/**
+ * Gets the upper bound of the expression, if the expression is guarded.
+ * An upper bound can only be found, if a guard phi node can be found, and no
+ * frontier phi exists for the phi node
+ */
+private float getGuardedUpperbound(Expr expr) {
+  exists(RangeSsaDefinition def, Variable v, VariableAccess guardVa, Expr guard, boolean branch |
+    v = expr.(VariableAccess).getTarget() and
+    expr = def.getAUse(v) and
+    def.isGuardPhi(guardVa, guard, branch) and
+    not def.hasFrontierPhi(v) and
+    guardVa.getTarget() = v and
```

Comment:
We must remember to look at the tuple counts for this predicate. It won't be easy to get this join order right, and I'm not even sure an optimal order exists.